Controls: Fix displaying as object instead of select for optional union types

tanujbhaud opened the pull request:
## What
Partially fixes #12641

## Why
When TypeScript properties are optional (`prop?: Type`), TypeScript adds `| undefined` to the union type. The enum detection logic in the TypeScript converter was failing because `undefined` is not a literal type, causing the union to not be recognized as an enum.

This resulted in controls displaying as "object" type instead of "select" dropdown for optional properties with union types, regardless of the number of items in the union.

## How
The fix filters out `undefined` elements before checking if all remaining elements are literals. This allows optional unions to be correctly recognized as enums and display as select controls.

### Changes
- Modified `code/core/src/docs-tools/argTypes/convert/typescript/convert.ts`
- Filter `undefined` from union elements when checking for enum type
- Use only non-undefined elements for enum value array

### Example
Before this fix:
```tsx
interface Props {
  myProperty?: 'k1' | 'k2' | ... | 'k29';  // Shows as "object" control
}
```

After this fix:
```tsx
interface Props {
  myProperty?: 'k1' | 'k2' | ... | 'k29';  // Shows as "select" control
}
```
## How to test manually

  1. Create a sandbox: yarn task --task dev --template react-vite/default-ts
  2. Create a component with optional union type props:
  interface Props {
    variant?: 'primary' | 'secondary' | 'tertiary';
    size: 'small' | 'medium' | 'large';
    manyOptions?: 'k1' | 'k2' | 'k3' | 'k4' | 'k5';
  }
  3. Create a story for the component and open it in Storybook
  4. Verify the Controls panel shows:
    - variant (optional) as radio buttons — not "object"
    - size (required) as radio buttons
    - manyOptions (optional) as a select dropdown — not "object"
  5. Verify you can select options and the component updates
  6. Verify the reset button (↺) clears optional props back to undefined

## Checklist
  - Fix tested locally
  - Minimal, focused change
  - Backward compatible with required unions
  - Type errors resolved
  - Unit tests pass

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Better detection and generation of enums from TypeScript unions that mix literals and undefined.

* **Changes**
  * Union processing now ignores undefined values when deriving enums and treats literal members distinctly.
  * Parser expects union elements to be explicit (stricter union shape) and recognizes literal values separately for more accurate output.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#12641: Problem with storybook-controls of type "select"](https://oss.issuehunt.io/repos/54173593/issues/12641)
---
</details>
<!-- /Issuehunt content-->

tanujbhaud committed:
Fix controls displaying as object instead of select for optional union types

Fixes #12641

When TypeScript properties are optional (prop?: Type), TypeScript adds
| undefined to the union type. The enum detection logic was failing
because undefined is not a literal type, causing the union to not be
recognized as an enum.

This resulted in controls displaying as "object" type instead of "select"
dropdown for optional properties with union types, regardless of the
number of items in the union.

The fix filters out undefined elements before checking if all remaining
elements are literals, allowing optional unions to be correctly recognized
as enums and display as select controls.

Sidnioulz reviewed (commented):
Thanks for the contribution @tanujbhaud!

Did you test whether it's possible to clear the select once an option has been selected?

We have another Select implementation in `storybook/internal/components` which has a reset option. @ghengeveld I'm curious to see what you think. Should we switch Form.Select to the newer Select with a reset option?

Sidnioulz commented on code/core/src/docs-tools/argTypes/convert/typescript/convert.ts:
```diff
@@ -43,12 +43,14 @@ export const convert = (type: TSType): SBType | void => {
-      if (type.elements?.every((element) => element.name === 'literal')) {
+      // Filter out undefined from optional props when checking for enum
+      const nonUndefinedElements = type.elements?.filter((element) => element.name !== 'undefined');
+      if (nonUndefinedElements?.length > 0 && nonUndefinedElements.every((element) => element.name === 'literal')) {
```
Type errors here if `nonUndefinedElements` is undefined.

Sidnioulz reviewed (changes requested):
CI reports several type errors, and you stripped the part of the PR template requesting a manual testing procedure.

Please provide both automatic tests and a manual testing procedure we can replicate to validate the PR.

Thanks,

tanujbhaud committed:
Fix type errors and linting issues in optional union enum handling

- Add type guard functions (isLiteral, isUndefined) for better type narrowing
- Add 'undefined' to TSScalarType union in types.ts
- Make elements non-optional in TSCombinationType (unions always have elements)
- Wrap case block in braces to fix Biome linting error
- Use proper type assertion instead of @ts-expect-error

tanujbhaud committed:
Merge branch 'next' into fix/controls-optional-union-enum

tanujbhaud committed:
Fix TypeScript type errors: add TSLiteralType with value property

- Separate TSLiteralType from TSScalarType to include 'value' property
- Use proper TypeScript type guard syntax (type is X) for narrowing
- Filter with type guard to get properly typed literal elements
- Removes need for explicit type casting

Sidnioulz commented:
> Hi @Sidnioulz, thanks for the review!
> 
> I've addressed the feedback:
> 
> 1. **Clear/reset functionality** - Tested and confirmed working. The reset button (↺) in the toolbar clears
>    optional props back to undefined.
> 2. **Type errors** - Fixed by:
>    
>    * Adding `'undefined'` to `TSScalarType` in `types.ts` so it matches runtime inputs
>    * Making `elements` non-optional in `TSCombinationType` (unions/intersections always have elements), which
>      fixes the "nonUndefinedElements is undefined" issue
> 3. **Type guards instead of @ts-expect-error** - Added `isLiteral()` and `isUndefined()` type guard functions
>    to properly narrow types without casting or suppressing errors
> 4. **Manual testing procedure** - Added to the PR description
> 
> Let me know if anything else is needed!

Thanks @tanujbhaud! Looks good to me, let's now see how CI behaves. Some tests are known to be flaky so you might see us update your branch and re-run CI a couple of times.

Sidnioulz committed:
Merge branch 'next' into fix/controls-optional-union-enum

tanujbhaud commented:
@valentinpalkovic @Sidnioulz Hey, just wanted to follow up, I submitted the PR on IssueHunt. Does it need an approval on your end there?
